templates/common/graph/graph.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package graph
+
+import (
+	"cmp"
+	"fmt"
+	"slices"
+
+	"golang.org/x/exp/maps"
+)
+
+// A DAG is a directed graph. It is list of nodes, identified by their slice
+// index, each having 0 or more outgoing edges to other node indices. So
+// [][]int{{1}, {}} is a graph with two nodes, where node 0 has an outgoing edge
+// to node 1, and node 1 has no outgoing edges.
+//
+// The edge directions are interpreted as "node index N depends on node indices
+// [X,Y,Z]". So the returned topological sort will place X, Y, and Z somewhere
+// before N.
+type DAG [][]int
```

Comment:
I'm always hesitant returning a direct data structure like this. It might be better to have a struct with a private field instead.

```go
type DAG struct {
  fields [][]int
}
```

and then some public functions for accessing things.

Also, should this be `int64`?

Reply:
Re hiding the [][]int : sure, will do.

Re int64: the values in the slice are slice indices, which are ints, so no they shouldn't be int64.